react/features/base/devices/services/device-detect/ActiveDeviceDetector.js

```diff
@@ -0,0 +1,143 @@
+// @flow
+import EventEmitter from 'events';
+import { ACTIVE_DEVICE_DETECTED } from './Events';
+import logger from '../../logger';
+import JitsiMeetJS from '../../../lib-jitsi-meet';
+
+const JitsiTrackEvents = JitsiMeetJS.events.track;
+
+// If after 3000 ms the detector did not find any active devices consider that there aren't any usable ones available
+// i.e. audioLevel > 0.008
+const DETECTION_TIMEOUT = 3000;
+
+/**
+ * Detect active input devices based on their audio levels, currently this is very simplistic. It works by simply
+ * checking all monitored devices for TRACK_AUDIO_LEVEL_CHANGED if a device has a audio level > 0.008 ( 0.008 is
+ * no input from the perspective of a JitsiLocalTrack ), at which point it triggers a ACTIVE_DEVICE_DETECTED event.
+ * If there are no devices that meet that criteria for DETECTION_TIMEOUT an event with empty deviceLabel parameter
+ * will be triggered,
+ * signaling that no active device was detected.
+ * TODO Potentially improve the active device detection using rnnoise VAD scoring.
+ */
+export class ActiveDeviceDetector extends EventEmitter {
```

Comment:
You are correct. But the reason It's structured like that is because at a certain point, we want to have the option of cancelling the current operation (in no-audio-signal/middleware.js) so we can start another one. We could probably achieve this with something like cancelable promises or other approaches but this felt a bit more clean.